### DISCUSSION TITLE
Classic Sequence style dynamic view

### DISCUSSION THREAD

**Discussion by NearlyUnique** [likec4/likec4#1108] [Ideas]
Are there any thoughts (and in fact is it feasible) to have a dynamic view render using classic vertical columns for actors/entities and messages horizontal lines for messages

**Comment by MrCoder** [comment]
Like sequence diagram?

**Reply by NearlyUnique** [reply]
Sorry, yes. The Dynamic View is a "sort of" sequence diagram. I wanted to be specific

The same information can be rendered in both styles. For some systems where a static view is needed then the standard UML sequence diagram allows for a more information dense diagram

**Reply by MrCoder** [reply]
I would definitely be interested. Given my experience on ZenUML, I might be able to help :)

**Reply by davydkov** [reply]
Yes, I think it would be beneficial, but not yet planned.

I found interesting that [stripe sequence diagrams](https://docs.stripe.com/payments/checkout/how-checkout-works?payment-ui=stripe-hosted#lifecycle) also based on [xyflow](https://xyflow.com/):

<img width="802" alt="SCR-20241021-mwxf" src="https://github.com/user-attachments/assets/ea4db3c9-6de7-4da2-8ad5-ccfe4f337942">



@MrCoder Eager to try?

<details><summary>how to</summary>
<p>
You can use <a href="https://github.com/likec4/likec4/compare/main...elk-experiment">elk-experiment</a> branch as a quick start.  

I experimented there with ELK layouts, as you see in diff.

After [initial build](https://github.com/likec4/likec4?tab=readme-ov-file#development), run `yarn dev` from `apps/playground`. Hot-Reload is ON - changes in any package do not require restart, but sometimes better refresh the page (due to Monaco)

Playground has dynamic diagram example - http://localhost:5173/w/dynamic/index/
</p>
</details>

**Reply by MrCoder** [reply]
Hi @davydkov, I haven't checked out the branch, so please excuse any inaccuracies in my comments. 

xyflow appears to be a drawing tool, while C4model is more focused on modeling—see [modelling vs diagramming](https://icepanel.io/blog/2024-08-06-comparison-c4-modelling-vs-diagramming). ZenUML's DSL is specifically designed for [modeling](https://zenuml.com/docs/concepts-and-methodologies/the-power-of-modeling-with-zenuml/). As the author of ZenUML, I’m happy to assist with development if needed.

Here is the reproduced version of the above diagram using ZenUML:
![image](https://github.com/user-attachments/assets/fdc58986-e021-4456-ba9e-a1bee74691c9)

ZenUML is interactive - https://diagramly.ai/diagrams/a1a7cd7a-15f2-42a6-bd98-d0f0db9630d6

DSL is like below:

```
@Actor Client
@Server Server
@Service StripeAPI
@Browser StripeCheckout

Client -> Server: Send order information
Server -> StripeAPI: Create Checkout Session
@return StripeAPI -> Server: Return Checkout Session
Server -> StripeCheckout: Redirect customer to url from Checkout Session
StripeCheckout -> StripeCheckout: Customer completes payment
StripeCheckout -> Client: Customer returns to your application
StripeAPI -> Server: Handle fulfillment
```

You can paste it on to https://app.zenuml.com to test it out.

**Comment by philippsimon** [comment]
It would be really nice to have this as an option for a dynamic view.

**Comment by davydkov** [comment]
Hey @philippsimon @NearlyUnique @MrCoder

Dynamic view supports two variants: `diagram` and `sequence`.

Here is example - https://likec4.dev/dsl/views/dynamic/#variants

**Reply by NearlyUnique** [reply]
Fantastic work people!

**Reply by philippsimon** [reply]
It looks really great! Thanks a lot!!

I only saw that only leaf nodes are supported. Will that change in the near future. Often we start first with system interaction where systems will later have children and more detailed diagrams are created. In that case it would break the diagrams.
